iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -434,6 +466,90 @@ public override async Task OpenAsync(CancellationToken cancellationToken)
                 _handlerSemaphore?.Release();
             }
         }
+        public override async void SetRefreshesOn(CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, cancellationToken, nameof(SetRefreshesOn));
+
+            DateTime refreshesOn = GetRefreshesOn();
+            if (refreshesOn < DateTime.MaxValue)
+            {
+                await StartLoopAsync(refreshesOn, cancellationToken);
+            }
+
+            if (Logging.IsEnabled)
+                Logging.Exit(this, cancellationToken, nameof(SetRefreshesOn));
+        }
+
+        private async Task StartLoopAsync(DateTime refreshesOn, CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, refreshesOn, nameof(StartLoopAsync));
+
+            if (_loopCancellationTokenSource == null
+                || _refreshLoop == null)
+            {
+                await StopLoopAsync();
+            }
+
+            _loopCancellationTokenSource = new CancellationTokenSource();
+            _refreshLoop = RefreshLoopAsync(refreshesOn, cancellationToken);
+
+            if (Logging.IsEnabled)
+                Logging.Exit(this, refreshesOn, nameof(StartLoopAsync));
+        }
+
+        private async Task RefreshLoopAsync(DateTime refreshesOn, CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, refreshesOn, nameof(RefreshLoopAsync));
+
+            TimeSpan waitTime = refreshesOn - DateTime.UtcNow;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Info(this, refreshesOn, $"Before {nameof(RefreshLoopAsync)} with wait time {waitTime}.");
+
+                if (waitTime > TimeSpan.Zero)
+                {
+                    await Task.Delay(waitTime, cancellationToken).ConfigureAwait(false);
```

Comment:
We should consider what thread this is running on. If this throws because the cancellation token is signaled, then it could harm the process.

If you we just want this method to quit, then we can try/catch (OperationCanceledException) { return; }

Reply:
Or maybe around the entire method body, because there is at least one other call with a cancellation token.

Reply:
And put Logging.Exit in the finally block.